### ISSUE TITLE
Have a more universal Faust code ?

### ISSUE THREAD

**Issue by sletz** [brummer10/VintageTubeOverdrive.lv2#1]
Could [this code](https://github.com/brummer10/VintageTubeOverdrive.lv2/blob/main/vintagetubeoverdrive/vintagetubeoverdrive.dsp#L87) calling an external piece of C code be rewritten as pure Faust code, possibly using `waveform` ?

**Comment by brummer10** [comment]
Yes. I've implemented the 1-dimensional function table for linear interpolation and the table itself as waveform. 
You could now select in the source which implementation you would use. 
regards
hermann

**Comment by sletz** [comment]
Thanks ! I guess @micbuffa will be happy, this way he or his team will be able to compile the Faust code as a [WAM2 plugin](https://github.com/micbuffa/wam-community) !

**Comment by micbuffa** [comment]
Hmmm I tried commenting/uncommenting the line, but the fast only version does not work when compiled in the faust IDE. It makes a strange noise then nothing. Input signal is not being processed...

**Comment by sletz** [comment]
Two problems detected with` interp-tracer` tool (see https://faustdoc.grame.fr/tutorials/debugging/).

`interp-tracer -trace 4 -input -ct 0  -sts  vintagetubeoverdrive.dsp ` gives:

```
-------- Interpreter crash trace start --------
assertLoadRealHeap array: fRealHeapSize 463 index 211 size 200 value 0.0225351 name ftbl0mydspSIG0
Stack [Int: 14] [REAL: 211,000000]
opcode 3 kLoadInt int 0 real 0 offset1 11 offset2 0 name iTemp3
Stack [Int: 211] [REAL: 211,000000]
opcode 32 kAddReal int 0 real 0 offset1 -1 offset2 -1
Stack [Int: 211] [REAL: -13526,411133]
opcode 2 kLoadReal int 0 real 0 offset1 460 offset2 0 name fTemp4
Stack [Int: 211] [REAL: 1,000000]
opcode 34 kSubReal int 0 real 0 offset1 -1 offset2 -1
Stack [Int: 211] [REAL: -13526,411133]
opcode 0 kRealValue int 0 real 1 offset1 -1 offset2 -1
Stack [Int: 211] [REAL: 1,488456]
opcode 2 kLoadReal int 0 real 0 offset1 459 offset2 0 name fTemp2
Stack [Int: 211] [REAL: 211,000000]
opcode 6 kStoreReal int 0 real 0 offset1 460 offset2 0 name fTemp4
Stack [Int: 211] [REAL: 1,488456]
opcode 26 kCastReal int 0 real 0 offset1 -1 offset2 -1
-------- Interpreter crash trace end --------
```

and `interp-tracer -trace 4 -input -ct 1  -sts  vintagetubeoverdrive.dsp` gives:

```
-------- Interpreter 'CastIntOverflow' trace start --------
opcode 27 kCastInt int 0 real 0 offset1 -1 offset2 -1
Stack [Int: 3] [REAL: 16086615,000000]
opcode 2 kLoadReal int 0 real 0 offset1 459 offset2 0 name fTemp2
Stack [Int: 3] [REAL: 2286599680,000000]
opcode 6 kStoreReal int 0 real 0 offset1 459 offset2 0 name fTemp2
Stack [Int: 3] [REAL: 16086615,000000]
opcode 36 kMultReal int 0 real 0 offset1 -1 offset2 -1
Stack [Int: 3] [REAL: 442121191424,000000]
opcode 190 kAbsf int 0 real 0 offset1 -1 offset2 -1
Stack [Int: 3] [REAL: 442121191424,000000]
opcode 2 kLoadReal int 0 real 0 offset1 458 offset2 0 name fTemp1
Stack [Int: 3] [REAL: 1,285143]
opcode 0 kRealValue int 0 real 142.143 offset1 -1 offset2 -1
Stack [Int: 3] [REAL: -16086615,000000]
opcode 6 kStoreReal int 0 real 0 offset1 458 offset2 0 name fTemp1
Stack [Int: 3] [REAL: 1,285143]
-------- Interpreter 'CastIntOverflow' trace end --------

```
Not that Faust latest master-dev branch has to be used.

**Comment by brummer10** [comment]
The issue is the first iir filter (p1). The faust IDE didn't cope well with higher order iir filters so we need to split it into lower order filters. I've update the dsp file now and verified that it works in the IDE.
regards
hermann

**Comment by sletz** [comment]
Thanks for the fix, working now ! (but see next message)

What do you mean by "The faust IDE didn't cope well with higher order iir filters" ? AFAICS this has nothing with the Faust IDE by itself, and the problem will appear with any possible compilation target (again I see : "reading outside tables" and "to big float cannot be converted to an integer" issues).

**Comment by sletz** [comment]
BTW there is still an "reading outside tables" issue with the new version.

Doing `interp-tracer -trace 4 -sts -input -ct 0 vintagetubeoverdrive.dsp` with the waveform version gives: 

```
-------- Interpreter crash trace start --------
assertLoadRealHeap array: fRealHeapSize 460 index 310 size 200 value 0.314014 name ftbl0mydspSIG0
Stack [Int: 8] [REAL: 310,000000]
opcode 3 kLoadInt int 0 real 0 offset1 12 offset2 0 name iTemp3
Stack [Int: 310] [REAL: 310,000000]
opcode 32 kAddReal int 0 real 0 offset1 -1 offset2 -1
Stack [Int: 310] [REAL: 141,058258]
opcode 2 kLoadReal int 0 real 0 offset1 457 offset2 0 name fTemp4
Stack [Int: 310] [REAL: 1,000000]
opcode 34 kSubReal int 0 real 0 offset1 -1 offset2 -1
Stack [Int: 310] [REAL: 141,058258]
opcode 0 kRealValue int 0 real 1 offset1 -1 offset2 -1
Stack [Int: 310] [REAL: 2,182000]
opcode 2 kLoadReal int 0 real 0 offset1 456 offset2 0 name fTemp2
Stack [Int: 310] [REAL: 310,000000]
opcode 6 kStoreReal int 0 real 0 offset1 457 offset2 0 name fTemp4
Stack [Int: 310] [REAL: 2,182000]
opcode 26 kCastReal int 0 real 0 offset1 -1 offset2 -1
-------- Interpreter crash trace end --------
```

This happen when the input is fed with `no.noise`.

Using  the`interp-tracer ` tool help a lot finding problems here !

**Comment by brummer10** [comment]
> What do you mean by "The faust IDE didn't cope well with higher order iir filters" ?

I've noticed that often with iir filters. Working well when used in local faust with double precision, but fail when using in the faust IDE. Lowering the filter order helps in any case, so far what I've tested.


> BTW there is still an "reading outside tables" issue with the new version.

I wonder how that could be, as the interpolation read pointer is strictly bounded to the table size here:

```
//-- Interpolate value between i and i+1 in table with fractal coefficient f.
interpolation(table, size, f, i) = select2(i<0,select2(i>size-2,
    rtable(table, i)*(1-f) + rtable(table,i+1)*f, rtable(table, size-1)),
    rtable(table, 0));

```
so even when the input value is to high, the read index is clamped.

**Comment by brummer10** [comment]
maybe adding a additional guard to linindex may avoid the "to big float cannot be converted to an integer"

linindex(step, x) = ma.fabs(x) * step, 200.0 : min;

**Comment by sletz** [comment]
> //-- Interpolate value between i and i+1 in table with fractal coefficient f.
> interpolation(table, size, f, i) = select2(i<0,select2(i>size-2,
>     rtable(table, i)*(1-f) + rtable(table,i+1)*f, rtable(table, size-1)),
>     rtable(table, 0));

Right. I was following this https://faustdoc.grame.fr/manual/faq/#does-select2-behaves-as-a-standard-cc-like-if in a bit too strict way, that is forcing both branches of the `select2` to be computed using the `-sts` option. But in practice, this does never happens. So if I debug now with:

`interp-tracer -trace 4  -input -ct 0 vintagetubeoverdrive.dsp `, I don't see problem anymore.

Note that the following code is possibly better, forcing the `i` index to stay [0..size-2] range before using it? (even if not the exact same semantic as your version that return `table[0]` when` i < 0` and  `table[size-1]` when` i > size-2`, but maybe still correct?) 

```
interpolation(table, size, f, i) = rtable(table, ci)*(1-f) + rtable(table, ci+1)*f
with {
    // Constraint i in [0..size-2] range
    ci = max(0, min(size-2, i));
};
```

**Comment by brummer10** [comment]
> Note that the following code is possibly better, forcing the `i` index to stay [0..size-2] range before using it?

Yes, it is. Nice. I'll switch to using that for future projects. Thanks for that.

**Comment by brummer10** [comment]
@micbuffa I've tried it now with the last pushed version as wasm plug, and it works as expected. (Pushed a trail version but forgot the link for the pedal board constructor to check it out with a following amp sim)

**Comment by micbuffa** [comment]
Indeed, it sounds good! I tried it rapidly. I will try to make it a good looking plugin tomorrow. I also have to check it guitar in hands first, sometimes some faust parameter ranges need to be a little adjusted when going faust -> webaudiomodules. We already ported some plugins from the guitarix team to WAMs (check in https://github.com/boourns/wam-community), we cited original source code authors in plugin descriptors...
